internal/command/jsonformat/state.go

```diff
@@ -0,0 +1,124 @@
+	"github.com/hashicorp/terraform/internal/command/jsonstate"
+)
+
+type State struct {
+	StateFormatVersion string                      `json:"state_format_version"`
+	RootModule         jsonstate.Module            `json:"root"`
+	RootModuleOutputs  map[string]jsonstate.Output `json:"root_module_outputs"`
+
+	ProviderFormatVersion string                            `json:"provider_format_version"`
+	ProviderSchemas       map[string]*jsonprovider.Provider `json:"provider_schemas"`
+}
+
+func (state State) Empty() bool {
+	return len(state.RootModuleOutputs) == 0 && len(state.RootModule.Resources) == 0 && len(state.RootModule.ChildModules) == 0
+}
+
+func (state State) GetSchema(resource jsonstate.Resource) *jsonprovider.Schema {
+	switch resource.Mode {
+	case jsonstate.ManagedResourceMode:
+		return state.ProviderSchemas[resource.ProviderName].ResourceSchemas[resource.Type]
+	case jsonstate.DataResourceMode:
+		return state.ProviderSchemas[resource.ProviderName].DataSourceSchemas[resource.Type]
+	default:
+		panic("found unrecognized resource mode: " + resource.Mode)
+	}
+}
+
+func (state State) renderHumanStateModule(renderer Renderer, module jsonstate.Module, opts computed.RenderHumanOpts, first bool) {
+	// Sort the resources in the module first, for consistent output.
+	var resources []jsonstate.Resource
+	resources = append(resources, module.Resources...)
+	sort.Slice(resources, func(i, j int) bool {
+		left := resources[i]
+		right := resources[j]
+
+		if left.Mode != right.Mode {
+			return left.Mode == jsonstate.DataResourceMode
+		}
+
+		if left.Address != right.Address {
+			return left.Address < right.Address
+		}
+
+		// Everything else being equal, we'll sort by deposed.
+		return left.DeposedKey < right.DeposedKey
+	})
+
+	if len(resources) > 0 && !first {
+		renderer.Streams.Println()
+	}
+
+	for _, resource := range resources {
+
+		if !first {
+			renderer.Streams.Println()
+		}
+
+		if first {
+			first = false
+		}
+
+		if len(resource.DeposedKey) > 0 {
+			renderer.Streams.Printf("# %s: (deposed object %s)", resource.Address, resource.DeposedKey)
+		} else if resource.Tainted {
+			renderer.Streams.Printf("# %s: (tainted)", resource.Address)
+		} else {
+			renderer.Streams.Printf("# %s:", resource.Address)
+		}
+
+		renderer.Streams.Println()
+
+		schema := state.GetSchema(resource)
+		switch resource.Mode {
+		case jsonstate.ManagedResourceMode:
+			renderer.Streams.Printf("resource %q %q %s", resource.Type, resource.Name, differ.FromJsonResource(resource).ComputeDiffForBlock(schema.Block).RenderHuman(0, opts))
+		case jsonstate.DataResourceMode:
+			renderer.Streams.Printf("data %q %q %s", resource.Type, resource.Name, differ.FromJsonResource(resource).ComputeDiffForBlock(schema.Block).RenderHuman(0, opts))
+		default:
+			panic("found unrecognized resource mode: " + resource.Mode)
+		}
+
+		renderer.Streams.Println()
+	}
+
+	// Render depth-first, so we show children first.
```

Comment:
I don't understand the "so we show children first" part of this comment. At first I thought this block should be before the loop over resources in this module, but I think the behaviour we have is reasonable, it just doesn't seem to quite line up with the comment. Maybe I'm missing something.

Reply:
We don't want to show the children first. This is left over from confusion caused by the old approach.

Previously, the order of the modules was not consistent and AFAICT fairly random. And I was doing diff testing, and noticed in one run that my order was different and thought it was because I had root first and children after. I rearranged the order as described by you and then re-ran my checks and got a different result. So I decided that it makes more sense to show the root first as I had previously, and that we would now have a consistent ordering but I never went back and removed this comment 😓 

I've removed the comment now, and since we're no longer making claims on the ordering I don't think a test is needed.

Reply:
I'm lacking some context here so this is just "FYI this exists" rather than any assumption it would be appropriate to use here, but I wanted to note that we have [`addrs.AbsResourceInstance.Less`](https://pkg.go.dev/github.com/hashicorp/terraform/internal/addrs#AbsResourceInstance.Less) to define our "default ordering" for sets of resource instances where we don't have any other specific order to use.

That `Less` function takes into account various details, such as sorting first by module membership and then by resource address, and making sure that we use numeric ordering rather than string-based ordering for the integer instance keys caused by the `count` feature.

It looks like this code is currently using a locally-implemented order, I guess in part because it needs to take into account the possibility of "deposed objects" which are outside of what `AbsResourceInstance` can describe. But perhaps we could use the canonical ordering for any pair of addresses that have differences beyond the deposed key.

Reply:
Thanks for this Martin, that's some good context.

I'll leave this as is for now, because I think a better approach if we wanted to do the sorting properly would be to do the sorting in the json packages that generate the JSON structs themselves, and then the renderer can just print the things without sorting. I'll have a look at this in a separate PR.

For this PR, I think it's more important just to make sure the output is consistent and then we don't have to worry about converting back into the Terraform structs. If we do the sorting before we marshal it into JSON then it's already in the Terraform structs.